plugins/woocommerce-blocks/assets/js/blocks/product-gallery/frontend.tsx

```diff
@@ -226,6 +171,55 @@ const productGallery = {
 					viewAllImagesElement;
 			}
 		},
+		onThumbnailKeyDown: ( event: KeyboardEvent ) => {
+			if (
+				event.code === 'Enter' ||
+				event.code === 'Space' ||
+				event.code === 'NumpadEnter'
+			) {
+				if ( event.code === 'Space' ) {
+					event.preventDefault();
+				}
+				productGallery.actions.selectImage();
+			}
+		},
+		onDialogKeyDown: ( event: KeyboardEvent ) => {
+			if ( event.code === 'Escape' ) {
+				actions.closeDialog();
+			}
+		},
+		openDialog: () => {
+			const context = getContext();
+			context.isDialogOpen = true;
+			const triggerElement = getElement()?.ref as HTMLElement;
+			if ( triggerElement ) {
+				context.elementThatTriggeredDialogOpening = triggerElement;
+			}
+			document.body.classList.add(
+				'wc-block-product-gallery-dialog-open'
+			);
+		},
+		closeDialog: () => {
+			// eslint-disable-next-line @wordpress/no-unused-vars-before-return
+			const context = getContext();
+			const clickedElement = getElement()?.ref as HTMLElement;
+			if ( clickedElement && clickedElement.tagName === 'IMG' ) {
+				// If the dialog is closed by clicking on an image, we need to select the image that was clicked.
+				const imgContext =
+					clickedElement.getAttribute( 'data-wc-context' );
+				if ( ! imgContext ) {
+					return null;
+				}
+				const imageId = JSON.parse( imgContext ).imageId;
+				context.selectedImageNumber =
+					getImageIndex( context, imageId ) + 1;
+				selectImage( getContext(), 'current' );
+			}
```

Comment:
As we discussed elsewhere, I kind of don't like this logic in practice.

1. IMO we _should not_ close the dialog on image click hence this logic could be removed
2. IMO we _should_ close the dialog on overlay click. If that's ok with you, you could include this in this PR or create an issue (should be fairly easy, attaching the on click directive to dialog 🤞 )